Done DLPAR rework for dedicated and shared mode.

SamirMulani opened the pull request:
Below code cahages are done,
-> Added support for avocado ssh library.
-> Added support to configure the DLPAR operation using .yaml file and generate tests.cfg file. -> The pexpect library's method of executing commands has been removed. -> Done code cleanup.
-> Added new README.md

Signed-off-by: Samir <samir@linux.vnet.ibm.com>

abdhaleegit commented:
@SamirMulani Could you please paste here the overall flow , may be block diagram of dlpar flow and also what different dlpar it supports ?

abdhaleegit reviewed (changes requested):
Logs please .. , also its a huge change, please split this into mulitple commits

SamirMulani commented:
> @SamirMulani Could you please paste here the overall flow , may be block diagram of dlpar flow and also what different dlpar it supports ?

**Dlpar Implementation details:**

1. Added support for avocado ssh library.
-> To execute the HMC commands in the old implementation, we used pexpect's python library for ssh operations. In the new implementation, we are using avocado's ssh library instead.
-> The implementation code for the pexpect way of communicatinon has been removed from dlpar_api/api.py.
**Changed files:**
1.1 dlpar/dlpar_api/api.py
1.2 dlpar/cpu_unit.py
1.3 dlpar/dedicated_cpu.py
1.4 dlpar/mem.py

2. Added support to configure the DLPAR operation using .yaml file and generate tests.cfg file.
-> The old implementation required maintaining a single test.cfg file for dedicated and shared mode operation. Hence to avoid maintaining one extra config file, we removed the test.cfg file from the config folder. Added support to configure the dlpar operation via a .yaml file and to generate the test.cfg file dynamically based on that .yaml file.
**Changed files:**
2.1 dlpar/dlpar_main.py

 
3. Added new README.md.
**Changed files:**
3.1 dlpar/README.md

**Dlpar-Operation supports:**
1. Dedicated mode cpu
-> ADD
-> REMOVE
-> MOVE

2. Dedicated mode mem
-> ADD
-> REMOVE
-> MOVE

3. Shared mode cpu
-> ADD
-> REMOVE
-> MOVE

4. Shared mode mem
-> ADD
-> REMOVE
-> MOVE


![DLPAR](https://user-images.githubusercontent.com/99895065/218387628-b1224239-2147-4f42-a418-eeeb01c823e4.jpeg)

SamirMulani commented:
> Logs please .. , also its a huge change, please split this into mulitple commits

```
[root]# avocado run dlpar_main.py -m dlpar_main.py.data/dlpar.yaml --max-parallel-tasks=1
JOB ID     : e3cad3d51de63503f71d6c79a0ddd1fbf4492fe7
JOB LOG    : /Workspace/AV/avocado-fvt-wrapper/results/job-2023-01-19T23.52-e3cad3d/job.log
 (1/2) dlpar_main.py:DlparTests.test_dlpar;run-config-lpar_mode-dedicated-test_type-cpu_dlpar-d527: STARTED
 (1/2) dlpar_main.py:DlparTests.test_dlpar;run-config-lpar_mode-dedicated-test_type-cpu_dlpar-d527: PASS (321.45 s)
 (2/2) dlpar_main.py:DlparTests.test_dlpar;run-config-lpar_mode-dedicated-test_type-mem_dlpar-355b: STARTED
 (2/2) dlpar_main.py:DlparTests.test_dlpar;run-config-lpar_mode-dedicated-test_type-mem_dlpar-355b: PASS (64.11 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /Workspace/AV/avocado-fvt-wrapper/results/job-2023-01-19T23.52-e3cad3d/results.html
JOB TIME   : 387.95 s
```

abdhaleegit commented on dlpar/dlpar_main.py:
```diff
@@ -55,105 +62,78 @@ def run_cmd(self, test_cmd, dlpar_type_flag=""):
+            Mem_obj.mem_add()
+            Mem_obj.mem_rem()
+            Mem_obj.mem_move()
+            # Mem_obj.mem_mix_ope()
```
so the same test will also cover shared scenarios ?

PraveenPenguin replied:
mem does not have shared concept

abdhaleegit reviewed (changes requested):
@SamirMulani Thank you for rework, the code looks more readable and structured now.. all looks good except few minor comments

abdhaleegit commented on dlpar/README.md:
```diff
@@ -13,68 +13,31 @@ OR
+hmc_user:        <hmc default user, ex: hscroot>
+hmc_passwd:      <hscroot password>
+target_lpar_name: <fully qualified  lpar name>
+target_partition: <partition name, should be same as lpar name>
```
rename to lpar_hostname

abdhaleegit replied:
the statement is confusing

SamirMulani replied:
No Both are different. It is addressed in a new commit with a unique name.

abdhaleegit commented on dlpar/dlpar_main.py.data/dlpar.yaml:
```diff
@@ -1,20 +1,20 @@
+cfg_cpu_per_proc: 8
+hmc_partition:
```
hmc_manageSystem

abdhaleegit commented on dlpar/dlpar_main.py.data/dlpar.yaml:
```diff
@@ -1,20 +1,20 @@
-         mem_dlpar:
-             test_case: 'mem'
-         
+            lp_mode: 'dedicated'
```
separate shared and dedicated

sacsant commented on dlpar/dlpar_api/api.py:
```diff
@@ -50,90 +47,47 @@ def __str__(self):
+            self.user = config_payload.get('hmc_user')
+            self.passwd = config_payload.get('hmc_passwd')
+            self.partition = config_payload.get('hmc_partition')
+            self.machine = config_payload.get('hmc_partition')
```
machine and partition are assigned same value. This seems incorrect. Can you check?

SamirMulani replied:
Yes, Bothe variables are the same and are needed in multiple places.

sacsant commented on dlpar/dlpar_api/api.py:
```diff
@@ -204,46 +154,260 @@ def check_log(self, message, code, die=True):
+                ' -p "' + linux_machine[0].partition + '"' + \
+                ' -t "' + linux_machine[1].partition + '"' + ' -w 0 '
+        else:
+            self.log.error("Envalid DLPAR flag")
```
Typo, Envalid --> Invalid

sacsant commented on dlpar/dlpar_api/api.py:
```diff
@@ -254,20 +418,37 @@ def get_cpu_option(self, linux_machine, option):
+        memory = 0
+        p_cmd = "cat /proc/meminfo | grep MemTotal | awk -F' ' '{ print $2 }'"
+        self.log.info('Checking memory under linux on %s' % machine.partition)
+        self.log.debug('Checking memory under linux on %s' % machine.partition)
```
One of the logging statement can be removed.

sacsant commented on dlpar/dlpar_api/api.py:
```diff
@@ -254,20 +418,37 @@ def get_cpu_option(self, linux_machine, option):
+        Check on a linux client how much physical memory we have.
+        """
+        memory = 0
+        p_cmd = "cat /proc/meminfo | grep MemTotal | awk -F' ' '{ print $2 }'"
```
What is the need for this statement? I see the p_cmd value is overwritten in the try block below. Only one statement should suffice.

SamirMulani replied:
This line is not needed, Removed in the latest commit.

sacsant commented on dlpar/dlpar_api/api.py:
```diff
@@ -304,7 +485,9 @@ def linux_check_rm_cpu(self, linux_machine, quantity_before,
-        cmd = linux_machine.sshcnx.run_command(c_cmd).decode()
+        cmd = process.run(c_cmd).stdout()
+        self.log.debug(cmd)
+        # cmd = linux_machine.sshcnx.cmd(c_cmd).decode()
```
any reason to keep this statement? If not can you remove it?

SamirMulani replied:
This line is not needed, Removed in the latest commit.

sacsant commented on dlpar/dlpar_api/api.py:
```diff
@@ -451,3 +629,552 @@ def set_virtual_proc_and_proc_units(self, linux_machine, ideal_procs,
+        self.quant_to_test = 2
+        self.quant_to_test = config_payload.get('ded_quantity_to_test')
+        self.sleep_time = 60
+        self.sleep_time = config_payload.get('sleep_time')
```
These variables are assigned values in succession. The first assignment is unnecessary.

SamirMulani replied:
The above changes are addressed in the latest commit

sacsant commented on dlpar/dlpar_api/api.py:
```diff
@@ -204,46 +154,260 @@ def check_log(self, message, code, die=True):
+            if not self.log.check_log(a_msg, a_condition, False):
+                e_msg = 'Error adding %d dedicated cpus to partition %s.' % \
+                        (quantity, linux_machine.partition)
+
```
Remove the blank line.

sacsant commented on dlpar/dlpar_api/api.py:
```diff
@@ -204,46 +154,260 @@ def check_log(self, message, code, die=True):
+                raise TestException(e_msg)
+
+        elif flag == "m":
+
```
same here.

sacsant commented on dlpar/dlpar_api/api.py:
```diff
@@ -451,3 +629,552 @@ def set_virtual_proc_and_proc_units(self, linux_machine, ideal_procs,
+        self.quant_to_test = 1024
+        self.quant_to_test = config_payload.get('mem_quantity_to_test')
+        self.sleep_time = 60
+        self.sleep_time = config_payload.get('sleep_time')
```
Values are overwritten by consecutive assignments. One of the assignment should be removed.

SamirMulani replied:
The above changes are addressed in the latest commit

SamirMulani commented:
Run Logs for DLPAR Dedicated and Shared mode,

**Dedicated mode Logs:**
```
[root@ltcden10-lp7 dlpar]# avocado run dlpar_main.py -m dlpar_main.py.data/dlpar.yaml --max-parallel-tasks=1
JOB ID     : a3728a04f70613a9359a34b3840c5faaceaf9d07
JOB LOG    : /Workspace/AV/avocado-fvt-wrapper/results/job-2023-03-01T12.57-a3728a0/job.log
 (1/2) dlpar_main.py:DlparTests.test_cpu_dlpar;run-config-lpar_mode-dedicated-c24d: STARTED
 (1/2) dlpar_main.py:DlparTests.test_cpu_dlpar;run-config-lpar_mode-dedicated-c24d: PASS (268.87 s)
 (2/2) dlpar_main.py:DlparTests.test_mem_dlpar;run-config-lpar_mode-dedicated-c24d: STARTED
 (2/2) dlpar_main.py:DlparTests.test_mem_dlpar;run-config-lpar_mode-dedicated-c24d: PASS (192.31 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /Workspace/AV/avocado-fvt-wrapper/results/job-2023-03-01T12.57-a3728a0/results.html
JOB TIME   : 471.25 s
```

**Shared mode Logs**
```
 avocado run dlpar_main.py -m dlpar_main.py.data/dlpar.yaml --max-parallel-tasks=1
JOB ID     : a2573e2234017be05a1ea5aca15041148c31511b
JOB LOG    : /Workspace/AV/avocado-fvt-wrapper/results/job-2023-03-07T00.27-a2573e2/job.log
 (1/2) dlpar_main.py:DlparTests.test_cpu_dlpar;run-config-lpar_mode-shared-f54f: STARTED
 (1/2) dlpar_main.py:DlparTests.test_cpu_dlpar;run-config-lpar_mode-shared-f54f: PASS (384.48 s)
 (2/2) dlpar_main.py:DlparTests.test_mem_dlpar;run-config-lpar_mode-shared-f54f: STARTED
 (2/2) dlpar_main.py:DlparTests.test_mem_dlpar;run-config-lpar_mode-shared-f54f: PASS (193.38 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /Workspace/AV/avocado-fvt-wrapper/results/job-2023-03-07T00.27-a2573e2/results.html
JOB TIME   : 580.21 s
```

PraveenPenguin commented on dlpar/dlpar_main.py:
```diff
@@ -55,105 +62,78 @@ def run_cmd(self, test_cmd, dlpar_type_flag=""):
+                                                    .splitlines():
+            if component in line:
+                a = line.split(':')[-1].strip()
+                print(a)
```
@SamirMulani please address this comment

SamirMulani replied:
Removed in the latest commit.

abdhaleegit commented on dlpar/dlpar_main.py.data/dedicated.yaml:
```diff
@@ -0,0 +1,20 @@
+hmc_user:
+hmc_passwd:
+target_lpar_name:
+target_partition:
```
rename to target_lpar_hostname  and lpar_name for hmc lpar name

abdhaleegit reviewed (changes requested):
Please look into last 2 review comments

SamirMulani commented:
Please find the new pull request for the above DLPAR implementation.
https://github.com/avocado-framework-tests/avocado-misc-tests/pull/2578
